### PR TITLE
Restrict user data access in User query

### PR DESCRIPTION
**Pull request by NishantSinghhhhh**
What kind of change does this PR introduce?

Implement user access control for querying user data.

Issue Number: #2214

Did you add tests for your changes?

Yes

Snapshots/Videos:

If relevant, did you update the documentation?

No

Summary

Does this PR introduce a breaking change?

No

## Summary by CodeRabbit

- **Security Enhancements**
    - Improved user data access controls
    - Added more granular permission checks for user queries

- **Error Handling**
    - Enhanced `NotFoundError` with additional properties for detailed error information
    - Refined error handling for user profile access scenarios

- **Testing**
    - Added comprehensive test suite for user query resolver
    - Verified access control scenarios for different user roles
    - Included tests for unauthorized access and retrieval of user profiles
    - Ensured proper error handling for non-existent user IDs

### CODE REVIEW

**Comment by NishantSinghhhhh** [comment]
@palisadoes , can you please review this PR

**Comment by palisadoes** [comment]
Please fix the failing tests. We need to get the code coverage for your patch to 100% without the use of ignore statements

**Comment by palisadoes** [comment]
Please fix the failing tests. Our test code coverage system will fail if any of these conditions occur:

1. The overall code coverage drops below the target threshold of the repository
2. Any file in the pull request has code coverage levels below the repository threshold
3. The code does not comply with our linting check or other related formatting requirements 

Here are some steps to take:

1. Click on the `Details` link for the tests to get an explanation of the cause(s)
2. You may have to fix a test that is not related to this PR as in some rare cases we merge faulty test code. If this is the case, fix the test in this PR. Do not open another issue.

We do this to ensure the overall reliability of the code base is maintained. This is a prerequisite for getting your PR approved. Assigned reviewers regularly review the PR queue and tend to focus on PRs that are passing.

Additional notes:

1. Ensure that CodeRabbit.ai approves your work
1. Please fix any conflicting files

**Comment by NishantSinghhhhh** [comment]
> Please fix the failing tests. Our test code coverage system will fail if any of these conditions occur:
> 
> 1. The overall code coverage drops below the target threshold of the repository
> 2. Any file in the pull request has code coverage levels below the repository threshold
> 3. The code does not comply with our linting check or other related formatting requirements
> 
> Here are some steps to take:
> 
> 1. Click on the `Details` link for the tests to get an explanation of the cause(s)
> 2. You may have to fix a test that is not related to this PR as in some rare cases we merge faulty test code. If this is the case, fix the test in this PR. Do not open another issue.
> 
> We do this to ensure the overall reliability of the code base is maintained. This is a prerequisite for getting your PR approved. Assigned reviewers regularly review the PR queue and tend to focus on PRs that are passing.
> 
> Additional notes:
> 
> 1. Ensure that CodeRabbit.ai approves your work
> 2. Please fix any conflicting files

Okay working on it

**Comment by palisadoes** [comment]
Please add the appropriate tests so that the code coverage tests pass

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please add the appropriate tests so that the code coverage tests pass

**Comment by NishantSinghhhhh** [comment]
@palisadoes  , I am trying to add as many tests that I can think of but still that specific part is uncovered

**Comment by NishantSinghhhhh** [comment]
@palisadoes , I have added numerous test for the specific uncovered part in user.ts , but still the tests are failing , can you please tell me what should I be doing, or what such test should I be adding so that 

User.ts is Fully covered

**Comment by palisadoes** [comment]
Please get code rabbit to approve your changes. The code coverage may get fixed then.

**Comment by NishantSinghhhhh** [comment]
@palisadoes , I tried implementing code-rabbbits changes too , still no progress in the codecov tests

File: tests/resolvers/User/userAccess.spec.ts [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+  it("throws NotFoundError when the queried user does not exist", async () => {
+    const context = {
+      userId: new Types.ObjectId().toString(), // Current user ID
+    };
+
+    // Step 1: Mock User.exists to return true
+    const mockUserExists = vi.spyOn(User, "exists").mockReturnValueOnce({
+      exec: async () => true, // Current user exists
+    } as any);
```
Comment (palisadoes) [inline tests/resolvers/User/userAccess.spec.ts:336]: Please fix the GitHub action request `as any`

File: tests/resolvers/User/userAccess.spec.ts [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+    // Step 2: Mock User.findById
+    const mockUserFind = vi.spyOn(User, "findById").mockImplementation(() => {
+      return {
+        exec: async () => null, // Queried user does not exist
+      } as any;
```
Comment (palisadoes) [inline tests/resolvers/User/userAccess.spec.ts:342]: Please fix the GitHub action request `as any`

File: tests/resolvers/User/userAccess.spec.ts [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+  it("throws NotFoundError when User.exists check fails", async () => {
+    const context = {
+      userId: new Types.ObjectId().toString(), // Simulating a current user ID
+    };
+
+    // Mock User.exists to return false
+    vi.spyOn(User, "exists").mockReturnValueOnce({
+      exec: async () => false,
+    } as any);
```
Comment (palisadoes) [inline tests/resolvers/User/userAccess.spec.ts:381]: Please fix the GitHub action request `as any`

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. See comments. 
2. Please add additional tests to make the code coverage tests pass.

**Comment by palisadoes** [comment]
Here is the missing line in the code coverage
- https://app.codecov.io/gh/PalisadoesFoundation/talawa-api/pull/2797/blob/src/resolvers/Query/user.ts

**Comment by NishantSinghhhhh** [comment]
@palisadoes , Done

**Review by palisadoes — APPROVED** [review]

File: tests/resolvers/User/userAccess.spec.ts [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+}));
+
+type UserType = {
+  _id: mongoose.Types.ObjectId;
+  email: string;
+  firstName?: string;
+  lastName?: string;
+  image?: string | null;
+  organizationsBlockedBy: string[];
+  [key: string]: any;
```
Comment (palisadoes) [inline tests/resolvers/User/userAccess.spec.ts:34]: Please resolve this

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please fix the linting error

**Comment by palisadoes** [comment]
Please fix the linting error

image

File: tests/resolvers/User/userAccess.spec.ts (hunk 4 above)
Comment (NishantSinghhhhh) [reply tests/resolvers/User/userAccess.spec.ts:34]: okay

**Comment by NishantSinghhhhh** [comment]
@palisadoes , Done with the issue

**Review by palisadoes — APPROVED** [review]

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Please fix. 
    image
1. You resolved a CodeRabbit suggestion without changing it.

**Comment by NishantSinghhhhh** [comment]
@palisadoes , my current code is not having the type any in userAcees.spec.ts

[Screencast from 2024-12-31 03-23-31.webm](https://github.com/user-attachments/assets/a9a0c88a-39f9-4fb2-abbe-19aa9f067fd2)

**Comment by palisadoes** [comment]
@coderabbitai full review

**Review by palisadoes — APPROVED** [review]

**Pull request merged**